src/tools/idlc/src/descriptor.c

```diff
@@ -1548,42 +1553,53 @@ static int print_opcode(FILE *fp, const struct instruction *inst)
-    case DDS_OP_VAL_8BY: vec[len++] = " | DDS_OP_TYPE_8BY"; break;
-    case DDS_OP_VAL_STR: vec[len++] = " | DDS_OP_TYPE_STR"; break;
-    case DDS_OP_VAL_BST: vec[len++] = " | DDS_OP_TYPE_BST"; break;
-    case DDS_OP_VAL_SEQ: vec[len++] = " | DDS_OP_TYPE_SEQ"; break;
-    case DDS_OP_VAL_ARR: vec[len++] = " | DDS_OP_TYPE_ARR"; break;
-    case DDS_OP_VAL_UNI: vec[len++] = " | DDS_OP_TYPE_UNI"; break;
-    case DDS_OP_VAL_STU: vec[len++] = " | DDS_OP_TYPE_STU"; break;
-    case DDS_OP_VAL_ENU: vec[len++] = " | DDS_OP_TYPE_ENU"; break;
-    case DDS_OP_VAL_EXT: vec[len++] = " | DDS_OP_TYPE_EXT"; break;
+  if (opcode == DDS_OP_ADR || opcode == DDS_OP_JEQ4) {
+    if (inst->data.opcode.code & DDS_OP_FLAG_EXT)
+      vec[len++] = " | DDS_OP_FLAG_EXT";
+    switch (DDS_OP_TYPE(inst->data.opcode.code)) {
+      case DDS_OP_VAL_1BY: vec[len++] = " | DDS_OP_TYPE_1BY"; break;
+      case DDS_OP_VAL_2BY: vec[len++] = " | DDS_OP_TYPE_2BY"; break;
+      case DDS_OP_VAL_4BY: vec[len++] = " | DDS_OP_TYPE_4BY"; break;
+      case DDS_OP_VAL_8BY: vec[len++] = " | DDS_OP_TYPE_8BY"; break;
+      case DDS_OP_VAL_STR: vec[len++] = " | DDS_OP_TYPE_STR"; break;
+      case DDS_OP_VAL_BST: vec[len++] = " | DDS_OP_TYPE_BST"; break;
+      case DDS_OP_VAL_SEQ: vec[len++] = " | DDS_OP_TYPE_SEQ"; break;
+      case DDS_OP_VAL_ARR: vec[len++] = " | DDS_OP_TYPE_ARR"; break;
+      case DDS_OP_VAL_UNI: vec[len++] = " | DDS_OP_TYPE_UNI"; break;
+      case DDS_OP_VAL_STU: vec[len++] = " | DDS_OP_TYPE_STU"; break;
+      case DDS_OP_VAL_ENU: vec[len++] = " | DDS_OP_TYPE_ENU"; break;
+      case DDS_OP_VAL_EXT: vec[len++] = " | DDS_OP_TYPE_EXT"; break;
+    }
   }
 
-  /* FLAG_BASE to indicate inheritance in PLM list or EXT 'parent' field */
   if (opcode == DDS_OP_ADR) {
+    /* FLAG_BASE to indicate EXT 'parent' field */
     if (inst->data.opcode.code & DDS_OP_FLAG_BASE)
       vec[len++] = " | DDS_OP_FLAG_BASE";
+    /* Must-understand */
+    if (inst->data.opcode.code & DDS_OP_FLAG_MU)
+      vec[len++] = " | DDS_OP_FLAG_MU";
+    /* Optional */
     if (inst->data.opcode.code & DDS_OP_FLAG_OPT)
       vec[len++] = " | DDS_OP_FLAG_OPT";
   }
-  if (opcode == DDS_OP_PLM && (DDS_PLM_FLAGS(inst->data.opcode.code) & DDS_OP_FLAG_BASE))
-    vec[len++] = " | (DDS_OP_FLAG_BASE << 16)";
+  if (opcode == DDS_OP_PLM) {
+    /* FLAG_BASE to indicate inheritance in PLM list */
+    if (DDS_PLM_FLAGS(inst->data.opcode.code) & DDS_OP_FLAG_BASE)
+      vec[len++] = " | (DDS_OP_FLAG_BASE << 16)";
+    /* Must-understand flag for mutable types set on PLM instruction */
+    if (DDS_PLM_FLAGS(inst->data.opcode.code) & DDS_OP_FLAG_MU)
+      vec[len++] = " | (DDS_OP_FLAG_MU << 16)";
+  }
 
   if (opcode == DDS_OP_JEQ || opcode == DDS_OP_JEQ4 || opcode == DDS_OP_PLM) {
     /* lower 16 bits contain an offset */
     idl_snprintf(buf, sizeof(buf), " | %u", (uint16_t) DDS_OP_JUMP (inst->data.opcode.code));
     vec[len++] = buf;
   } else {
```

Comment:
If this is turned into an `else if (opcode == DDS_OP_ADR)` neither the assert below nor the `goto print`s in the first switch statement of the function appear to be needed any longer. It is a matter of taste whether removing the `goto print`s would improve things, but I suspect it would, simply by eliminating the question: "why are these treated differently?".

Reply:
Good point, the `goto` also bothered me. I've refactored this a bit in c741607